File: src/tools/diagnose_github_token.py
#!/usr/bin/env python3
"""
GitHub Token Diagnostic Tool
Helps diagnose GitHub token issues and permissions
"""

import os
import requests
import json
from datetime import datetime

def check_github_token():
    """Check GitHub token validity and permissions"""
    print("🔍 GitHub Token Diagnostic")
    print("=" * 30)
    
    # Check if token exists
    github_token = os.getenv('GITHUB_TOKEN')
    if not github_token:
        print("❌ GITHUB_TOKEN not found in environment variables")
        print("\n💡 To set your GitHub token:")
        print("   Windows PowerShell:")
        print("   $env:GITHUB_TOKEN='your_token_here'")
        print("   ")
        print("   Or create a .env file with:")
        print("   GITHUB_TOKEN=your_token_here")
        return False
    
    print(f"✅ GITHUB_TOKEN found: {github_token[:12]}...")
    
    # Test basic GitHub API access
    print("\n🔧 Testing Basic GitHub API Access...")
    try:
        headers = {
            "Authorization": f"Bearer {github_token}",
            "Accept": "application/vnd.github.v3+json"
        }
        
        response = requests.get("https://api.github.com/user", headers=headers)
        
        if response.status_code == 200:
            user_data = response.json()
            print(f"✅ GitHub API Access: OK")
            print(f"   User: {user_data.get('login', 'Unknown')}")
            print(f"   Name: {user_data.get('name', 'Not set')}")
        else:
            print(f"❌ GitHub API Access Failed: {response.status_code}")
            print(f"   Error: {response.text}")
            return False
            
    except Exception as e:
        print(f"❌ GitHub API test failed: {str(e)}")
        return False
    
    # Test GitHub Models API specifically
    print("\n🤖 Testing GitHub Models API Access...")
    try:
        headers = {
            "Authorization": f"Bearer {github_token}",
            "Content-Type": "application/json"
        }
        
        # Test with a simple request
        data = {
            "model": "gpt-4o-mini",
            "messages": [
                {
                    "role": "user", 
                    "content": "Hello, can you respond with 'GitHub Models API is working'?"
                }
            ],
            "max_tokens": 20,
            "temperature": 0.1
        }
        
        response = requests.post(
            "https://models.inference.ai.azure.com/chat/completions",
            json=data,
            headers=headers,
            timeout=10
        )
        
        if response.status_code == 200:
            result: str = response.json()
            content = result["choices"][0]["message"]["content"]
            print("✅ GitHub Models API: Working!")
            print(f"   Response: {content}")
            return True
        else:
            print(f"❌ GitHub Models API Error: {response.status_code}")
            try:
                error_data = response.json()
                print(f"   Error Details: {json.dumps(error_data, indent=2)}")
            except:
                print(f"   Raw Error: {response.text}")
            
            # Provide specific guidance based on error
            if response.status_code == 401:
                print("\n💡 Token Permission Issue:")
                print("   Your token may not have 'models' permission")
                print("   1. Go to: https://github.com/settings/personal-access-tokens/tokens")
                print("   2. Edit your token or create a new one")
                print("   3. Ensure 'models' permission is selected")
            elif response.status_code == 403:
                print("\n💡 Access Issue:")
                print("   GitHub Models might not be available for your account")
                print("   1. Check if you have GitHub Copilot subscription")
                print("   2. Verify GitHub Models is enabled for your organization")
            
            return False
            
    except Exception as e:
        print(f"❌ GitHub Models test failed: {str(e)}")
        return False

def show_setup_guide():
    """Show detailed setup guide"""
    print("\n📋 GitHub Token Setup Guide")
    print("=" * 30)
    print("""
🔧 Step 1: Create GitHub Token
   1. Go to: https://github.com/settings/personal-access-tokens/tokens
   2. Click "Generate new token"
   3. Select "Fine-grained personal access token" (recommended)
   4. Set expiration (90 days recommended)
   5. Select permissions:
      ✅ models (required for GitHub Models API)
      ✅ metadata (basic access)

🔧 Step 2: Set Environment Variable
   Windows PowerShell:
   $env:GITHUB_TOKEN='ghp_xxxxxxxxxxxx'
   
   Or create .env file:
   GITHUB_TOKEN=ghp_xxxxxxxxxxxx

🔧 Step 3: Verify Access
   Run this script again to verify setup

⚠️  Important Notes:
   - You need GitHub Copilot subscription
   - Token must have 'models' permission
   - Some organizations may restrict GitHub Models access
""")

def main():
    """Main diagnostic function"""
    print(f"🏦 GitHub Token Diagnostic - {datetime.now().strftime('%Y-%m-%d %H:%M:%S')}")
    print("=" * 60)
    
    success = check_github_token()
    
    if not success:
        show_setup_guide()
        print("\n❌ GitHub token setup needs attention")
        return False
    else:
        print("\n🎉 GitHub token is properly configured!")
        print("\n🚀 Your GitHub Copilot integration should work now")
        return True

if __name__ == "__main__":
    main()

File: scripts/diagnose-github-token.js
const https = require('https');
require('dotenv').config();

console.log('🔍 GitHub Token Diagnostic Tool');
console.log('================================');

const token = process.env.GITHUB_TOKEN;

if (!token) {
    console.log('❌ GITHUB_TOKEN not found in environment');
    process.exit(1);
}

console.log(`🔑 Token format: ${token.substring(0, 20)}...`);
console.log(`📏 Token length: ${token.length}`);
console.log(`✅ Token prefix: ${token.startsWith('github_pat_') ? 'Correct (github_pat_)' : 'Incorrect - should start with github_pat_'}`);

// Check if token has the right format
if (!token.startsWith('github_pat_')) {
    console.log('❌ Token format error: GitHub Personal Access Tokens should start with "github_pat_"');
    console.log('📋 Please generate a new token at: https://github.com/settings/tokens');
    process.exit(1);
}

// Test basic GitHub API access
console.log('\n🧪 Testing basic GitHub API access...');

const options = {
    hostname: 'api.github.com',
    path: '/user',
    method: 'GET',
    headers: {
        'Authorization': `Bearer ${token}`,
        'User-Agent': 'Flash-Loan-Arbitrage-System',
        'Accept': 'application/vnd.github.v3+json'
    }
};

const req = https.request(options, (res) => {
    let data = '';
    
    res.on('data', (chunk) => {
        data += chunk;
    });
    
    res.on('end', () => {
        console.log(`📊 Status Code: ${res.statusCode}`);
        
        if (res.statusCode === 200) {
            const user = JSON.parse(data);
            console.log(`✅ Token valid! User: ${user.login}`);
            console.log(`👤 Account type: ${user.type}`);
            console.log(`📅 Account created: ${user.created_at}`);
            
            // Now test GitHub Models API
            testGitHubModels();
        } else {
            console.log(`❌ GitHub API Error: ${res.statusCode}`);
            console.log(`📝 Response: ${data}`);
            
            if (res.statusCode === 401) {
                console.log('\n🔧 Troubleshooting steps:');
                console.log('1. Verify token is correctly copied from GitHub');
                console.log('2. Check token hasn\'t expired');
                console.log('3. Ensure token has required scopes');
                console.log('4. Generate new token at: https://github.com/settings/tokens');
            }
        }
    });
});

req.on('error', (err) => {
    console.error('❌ Request error:', err.message);
});

req.end();

function testGitHubModels() {
    console.log('\n🤖 Testing GitHub Models API access...');
    
    const modelOptions = {
        hostname: 'models.inference.ai.azure.com',
        path: '/chat/completions',
        method: 'POST',
        headers: {
            'Authorization': `Bearer ${token}`,
            'Content-Type': 'application/json',
            'User-Agent': 'Flash-Loan-Arbitrage-System'
        }
    };

    const requestData = JSON.stringify({
        messages: [
            {
                role: "user",
                content: "Hello! Can you confirm you're GitHub Copilot?"
            }
        ],
        model: "gpt-4o",
        max_tokens: 50
    });

    const modelReq = https.request(modelOptions, (res) => {
        let data = '';
        
        res.on('data', (chunk) => {
            data += chunk;
        });
        
        res.on('end', () => {
            console.log(`📊 Models API Status: ${res.statusCode}`);
            
            if (res.statusCode === 200) {
                const response = JSON.parse(data);
                console.log('✅ GitHub Models API working!');
                console.log(`🤖 Response: ${response.choices[0].message.content}`);
            } else {
                console.log(`❌ Models API Error: ${res.statusCode}`);
                console.log(`📝 Response: ${data}`);
                
                if (res.statusCode === 401) {
                    console.log('\n💡 GitHub Models might require:');
                    console.log('1. GitHub Copilot subscription');
                    console.log('2. Specific token scopes');
                    console.log('3. Organization access (if using organization account)');
                } else if (res.statusCode === 403) {
                    console.log('\n💡 Access forbidden - check:');
                    console.log('1. GitHub Copilot subscription is active');
                    console.log('2. Account has access to GitHub Models');
                    console.log('3. Token has required permissions');
                }
            }
        });
    });

    modelReq.on('error', (err) => {
        console.error('❌ Models API request error:', err.message);
    });

    modelReq.write(requestData);
    modelReq.end();
}
